### ISSUE TITLE
Questions on Thread prefix and route management

### ISSUE THREAD

**Issue by gjc13** [Abhayakara/mdnsresponder#1]
# Number of prefix supported
Currently only we can only publish one Thread prefix and this prefix is added by the function `ula_setup`. Shall we check whether there is an already-setup off-mesh prefix? 
Is there a plan to support advertising multiple prefixes and prefixes advertised from other nodes in the Thread network?

# Routing advertisement packet contents

## MTU

When sending ICMP RA packets, it seems that we're only sending MTU when the interface is the Thread interface: https://github.com/Abhayakara/mdnsresponder/blob/a28d5ee94aa112727937b6c28c496dae9f4ee43e/ServiceRegistration/route.c#L1462

However due to this line https://github.com/Abhayakara/mdnsresponder/blob/a28d5ee94aa112727937b6c28c496dae9f4ee43e/ServiceRegistration/route.c#L1407 this MTU option will never be sent.

Should we include the MTU in RA packets sent on other interfaces so that WiFi devices can be informed with the Thread MTU?

## Route information on ULA prefix

In this code section https://github.com/Abhayakara/mdnsresponder/blob/a28d5ee94aa112727937b6c28c496dae9f4ee43e/ServiceRegistration/route.c#L1557

We'll publish route info on the ULA::/48 prefix. Will this prefix be shared between the wifi and Thread devices? If such being the case, is it proper to advertise route for this prefix on the Thread border router?

# ICMP RA message handling

## Address configuration

In the `interface_prefix_configure` function:
https://github.com/Abhayakara/mdnsresponder/blob/a28d5ee94aa112727937b6c28c496dae9f4ee43e/ServiceRegistration/route.c#L1167

We will configure address `prefix:1/64` on the interface receiving ICMP RA message. Shall we have the kernel configure the address for us depending on the DHCP and SLAAC flag?

##  OpenThread routing to other networks
When receiving a new prefix on other interfaces, currently we are not forwarding this route information to the Thread network.

Shall we add a function call to `otBorderRouterAddRoute`?

**Comment by jwhui** [comment]
> Shall we add a function call to otBorderRouterAddRoute?

In a Thread network attached to a single infrastructure link, default route should be sufficient for routing from a Thread device to an off-mesh destination.

**Comment by Abhayakara** [comment]
# Number of prefix supported

The way the code currently works is that we look to see if there is already an off-mesh-routable prefix; if there is, we just publish that one to signal that we are a thread border router, rather than adding a second prefix, which would waste space in the network data and in SRP updates.

What you see in ula_setup is that each BR generates its own /48 ULA prefix using a good random number generator, and then saves that in stable storage for re-use. This means that each thread BR is in a position to set up an on-mesh or infrastructure prefix at any time, and can be assured that it will not conflict with a prefix in use on a Thread partition that is not visible to that BR.

If we were to use a common prefix amongst all border routers, we'd need some way to make sure that we never accidentally wound up with two thread partitions with the same off-mesh-routable on-link prefix, each of which might be injected into the site routing table resulting in ambiguous routes and unreliable routing.

If we decide to implement DUA, we will need a way to make this work; I would suggest routing the DUA to a NAT66 translator rather than actually using the DUA prefix on thread meshes, so I don't know of a use case for multiple off-mesh-routable prefixes on a single Thread mesh in normal use.

That said, there is a case where this can happen: if a mesh is partitioned, the BRs will automatically make sure that the new partition has an off-mesh-routable prefix. If that partition then heals, we need to keep the new prefix around for a while so that communication with devices on the former "new partition" is not interrupted. Currently that time period is an hour; it's not clear that this is the right number. This mechanism also causes some problems because during mesh reformation the partition ID can change multiple times, and the end state can be a stable mesh that requires only one prefix, so we don't want multiple prefixes per mesh. There is further work to do here.

# MTU

Publishing the MTU in an RA sent to the Thread network is obviously not useful since we don't send RAs to the Thread network. Currently I don't think we actually know the MTU of the infrastructure network, so we can't set it. I am open to change here if you think this is a problem we should address.

# Route information on ULA prefix

The SKIP_SLASH_48 code should always be compiled; when this is defined, we only publish a /64. I was experimenting here, and so there's some code that's wrapped in ifdefs that's part of a failed experiment and should be deleted.

# Address configuration

The Thread Border Router is, when it publishes an on-link prefix on infrastructure, doing so as a router. There is no DHCP server that can give us this address, and trying to use SLAAC to configure this prefix is problematic. There's no downside to using the ::1 address that I know of, hence that's what the code does.

Ideally, there is already IPv6 service on the link. When this is the case, we act as a host and accept the prefix that the network has offered, which means we use SLAAC or DHCP to get an address, and do not configure one directly.

# OpenThread routing to other networks

It's possible that two Thread BRs connected to the same Thread mesh might be connected to different adjacent infrastructure links. In this case, it would be ideal to add a route for each on-link prefix that's present on an adjacent infrastructure link to which the BR has direct reachability. In principle, if the infrastructure network is properly routed, this is not necessary, but I agree that it's probably good practice.

Thanks for your questions!

**Comment by gjc13** [comment]
Thanks for your answer. What's the future plan for code clean up and merging? Are we aiming to merge some of the code into OpenThread/ot-br-posix?

**Comment by Abhayakara** [comment]
I think it would make sense to merge route.c into ot-daemon. If we did that, we would still need communication between ot-daemon and srp-mdns-proxy, but not as much of it. The code is a bit of a giant pile right now, and it would be good to separate out functional blocks. E.g., tracking state of link is one block. Tracking state of Thread mesh is another block. Each of these no doubt subdivide into smaller blocks. Basically, systematizing the very ad-hoc solution that has evolved as I did this work.
